docs/reference/service-discovery.md

```diff
@@ -0,0 +1,263 @@
+
+```python
+class MyApp(asab.Application):
+	def __init__(self):
+		super.__init__(self)
+		# Initialize web server
+		self.add_module(asab.web.Module)
+		websvc = self.get_service("asab.WebService")
+		self.WebContainer = asab.web.WebContainer(websvc, "web")
+
+		# Initialize zookeeper
+		self.add_module(asab.zookeeper.Module)
+		self.ZooKeeperService = self.get_service("asab.ZooKeeperService")
+		self.ZooKeeperContainer = asab.zookeeper.ZooKeeperContainer(
+			self.ZooKeeperService, "zookeeper"
+		)
+
+		# Initialize ApiService
+		self.ASABApiService = asab.api.ApiService(self)
+		self.ASABApiService.initialize_web(self.WebContainer)
+		self.ASABApiService.initialize_zookeeper(self.ZooKeeperContainer)
+```
+
+### Configuration
+
+ZooKeeper configuration must be the same for all services in the cluster. Port can be set in the configuration.
+
+```ini title="myapp.conf"
+[web]
+listen=0.0.0.0 8090
+
+[zookeeper]
+servers=zookeeper-1:2181,zookeeper-2:2181,zookeeper-3:2181
+```
+
+### Environment variables
+
+The discovery is based on provided "ids". Use either a set of specific ids provided as environemntal variable, or create [custom discovery ids](#custom-discovery-ids) within the application.
+
+Pre-cast ids used for service discovery. Provide them as environment variables:
+
+| Variable | Description | Example |
+| --- | --- | --- |
+| `SERVICE_ID` | Identifier of the ASAB microservice. | my-app |
+| `INSTANCE_ID` | Identifier of a specific instance of the ASAB microservice. | my-app-1, my-app-2, ... |
+| `NODE_ID` | Identifier of a cluster node. | node-1, node-2, ... |
+
+!!! warning "Instance ID must be resolvable"
+
+	ASAB framework cannot set up networking. In order to enable service discovery, `INSTANCE_ID` of a service must be resolavable from the client. If `INSTANCE_ID` is missing, hostname is taken instead and then hostname must be resolvable.
+
+
+### Information advertised
+
+When all requirements of a server-side microservice are fullfilled, the service advertises into the consensus a unique information bound to its runtime.
+
+
+```json title="asab/run/ASABMyApplication.0000090098"
+{
+	"host": "server-name-1",
+	"appclass": "MyApp",
+	"node_id": "node-1",
+	"service_id": "my-app",
+	"instance_id": "my-app-1",
+	"launch_time": "2023-12-01T10:52:29.656397Z",
+	"process_id": 1,
+	"containerized": true,
+	"created_at": "2023-11-23T11:08:22.454248Z",
+	"version": "v23.47-alpha",
+	"CI_COMMIT_TAG": "v23.47-alpha",
+	"CI_COMMIT_REF_NAME": "v23.47-alpha",
+	"CI_COMMIT_SHA": "...",
+	"CI_COMMIT_TIMESTAMP": "2023-11-23T11:00:48+00:00",
+	"CI_JOB_ID": "...",
+	"CI_PIPELINE_CREATED_AT": "2023-11-23T11:06:40Z",
+	"CI_RUNNER_ID": "..",
+	"CI_RUNNER_EXECUTABLE_ARCH": "linux/amd64",
+	"web": [
+		[
+			"0.0.0.0",
+			8090
+		]
+	]
+}
+```
+
+## Client - Using Service Discovery
+
+As all services are being "server" and "client" at the same time, all requirements stated for server microservice are valid for the client one, as well.
```

Comment:
This is a bit fuzzy ... not sure what you mean by "requirements" here.

This can be solved by section "Requirements" that will be in the beginning and it will define all requirements for servers and clients.